### ISSUE TITLE
[Microsoft Generator CSharp] Refactor ApiTypes Properties into Expression / Statements Overrides

### ISSUE THREAD

**Issue by jorgerangel-msft** [Azure/autorest.csharp#4226]
Currently, the [ApiTypes](https://github.com/Azure/autorest.csharp/blob/feature/v3/src/AutoRest.CSharp/Common/Input/ApiTypes.cs#L20) class contains abstract properties for certain types, ie. `ResponseType`. This design should be refactored to be backed into expression and statement overrides instead, which extending plugins can override. In addition, ApiTypes should be combined with the ExensibleSnippets class, as there are several overlaps.

Original discussion: https://github.com/jorgerangel-msft/Microsoft.Generator.CSharp/pull/12#discussion_r1491485984 and https://github.com/jorgerangel-msft/Microsoft.Generator.CSharp/pull/13#discussion_r1496720672
```[tasklist]
### Tasks
- [ ] https://github.com/Azure/autorest.csharp/issues/4721
- [ ] [Microsoft Generator CSharp] Refactor response types: public abstract Type ResponseType { get; }         public abstract Type ResponseOfTType { get; }
```

**Comment by ArcturusZhang** [comment]
Semantically speaking, the responsibility of "getting what type we should have for a client response" should belong to the `TypeFactory` as its name suggests.
And for those giving a complicated snippets, we could put them into the Snippets extensions.

**Comment by m-nash** [comment]
I think its also fair to have methods on the TypeFactory which return an Expression.

For instance lets imagine we have a generic ChangeTrackingListExpression and instead of ApiTypes with generic snippets we have a method on TypeFactory

```c#
public virtual ChangeTrackingListExpression GetChangeTrackingList();
```

Now people can get and use that expression by its interface without needing to know the implementation or call into a separate place where we add extensions such as snippets.

I think haven't multiple places people need to go for these extensions is probably a bad thing.

**Comment by ArcturusZhang** [comment]
> I think its also fair to have methods on the TypeFactory which return an Expression.
> 
> For instance lets imagine we have a generic ChangeTrackingListExpression and instead of ApiTypes with generic snippets we have a method on TypeFactory
> 
> ```cs
> public virtual ChangeTrackingListExpression GetChangeTrackingList();
> ```
> 
> Now people can get and use that expression by its interface without needing to know the implementation or call into a separate place where we add extensions such as snippets.
> 
> I think haven't multiple places people need to go for these extensions is probably a bad thing.

I agree but this could lead to that we are putting everything into `TypeFactory` therefore we should add some organization for it so that it could have hierarchies.

**Comment by nisha-bhatia** [comment]
[#4720](https://github.com/Azure/autorest.csharp/issues/4720)

**Comment by JoshLove-msft** [comment]
I believe this has been done as part of the plugin abstractions that were added for the Azure plugin.
